src/Raven.Server/Documents/ETL/EtlProcess.cs

```diff
@@ -706,29 +654,26 @@ public void Run()
                                 EnsureThreadAllocationStats();
 
                                 using (context.OpenReadTransaction())
+                                using (var scope = new DisposableScope())
+                                using (var merged = new ExtractedItemsEnumerator<TExtracted>(stats))
                                 {
-                                    var transformations = ExtractTransform(EtlItemType.Document);
+                                    var nextEtag = loadLastProcessedEtag + 1;
+
+                                    Extract(context, merged, nextEtag, EtlItemType.Document, stats, scope);
 
                                     if (ShouldTrackCounters())
-                                        transformations.AddRange(ExtractTransform(EtlItemType.CounterGroup));
+                                        Extract(context, merged, nextEtag, EtlItemType.CounterGroup, stats, scope);
                                     
                                     if (ShouldTrackTimeSeries())
-                                        transformations.AddRange(ExtractTransform(EtlItemType.TimeSeries));
+                                        Extract(context, merged, nextEtag, EtlItemType.TimeSeries, stats, scope);
 
-                                    List<TTransformed> ExtractTransform(EtlItemType type)
-                                    {
-                                        var extracted = Extract(context, loadLastProcessedEtag + 1, type, stats);
-                                        return Transform(extracted, context, stats, state);
-                                    }
+                                    var transformations = Transform(merged, context, stats, state);
 
-                                    var successfulLoad = false;
-
-                                    if (transformations.Count > 0)
```

Comment:
In a case, there is no transformed item you still act as "didWork"

Reply:
yes, because there might be stuff that we skipped, but we need to update the etag.